dispatch-guard positional fallback deadlocks mixed-dependency milestones

## Problem

Auto-mode deadlocks when a milestone mixes explicit `depends` declarations with slices that have no dependencies. Slices with empty `depends` fall into the positional fallback path and get blocked by earlier incomplete slices — even when the blocking slice itself depends on a later slice.

**Error:** `Cannot dispatch execute-task M022/S06/T02: earlier slice M022/S05 is not complete.`

S06 declares no dependencies (`depends: —`). S05 declares `depends: [S02, S03, S07]`. S07 is positionally later than S05. This creates a deadlock: S06 is blocked by S05 (positional), S05 is blocked by S07 (explicit dep), and S06 can never proceed.

## Root Cause

**`dispatch-guard.js:93-112`** — The slice ordering guard has two code paths:

```javascript
// Line 93: dependency-aware path
if (targetSlice.depends.length > 0) {
    const sliceMap = new Map(slices.map((s) => [s.id, s]));
    for (const depId of targetSlice.depends) {
        const dep = sliceMap.get(depId);
        if (dep && !dep.done) {
            return `Cannot dispatch ${unitType} ${unitId}: dependency slice ${targetMid}/${depId} is not complete.`;
        }
    }
}
// Line 104: positional fallback
else {
    const targetIndex = slices.findIndex((slice) => slice.id === targetSid);
    const incomplete = slices
        .slice(0, targetIndex)
        .find((slice) => !slice.done);
    if (incomplete) {
        return `Cannot dispatch ${unitType} ${unitId}: earlier slice ${targetMid}/${incomplete.id} is not complete.`;
    }
}
```

When **any** slice in the milestone uses explicit `depends`, the milestone is operating in dependency-graph mode. But slices with empty `depends` still fall into the positional fallback (line 104), which assumes sequential ordering. This creates an inconsistency: the milestone is half dependency-graph, half positional.

## Expected Behavior

When a milestone contains **any** slice with explicit dependencies, all slices should use the dependency graph. Empty `depends` should mean "no dependencies" (immediately dispatchable), not "depends on all earlier slices."

**Suggested fix** at `dispatch-guard.js:104`:

```javascript
else {
    // Only apply positional fallback if NO slice in this milestone
    // uses explicit depends. Mixed milestones use the dep graph —
    // empty depends means "independent."
    const milestoneUsesExplicitDeps = slices.some(s => s.depends.length > 0);
    if (!milestoneUsesExplicitDeps) {
        const targetIndex = slices.findIndex((slice) => slice.id === targetSid);
        const incomplete = slices
            .slice(0, targetIndex)
            .find((slice) => !slice.done);
        if (incomplete) {
            return `Cannot dispatch ${unitType} ${unitId}: earlier slice ${targetMid}/${incomplete.id} is not complete.`;
        }
    }
}
```

## Environment
- GSD version: 2.70.1
- Model: claude-opus-4-6
- Unit: execute-task/M022/S06/T02

## Reproduction Context

Milestone M022 has 7 slices. S01-S04 use explicit `depends`. S05 depends on `[S02, S03, S07]` (S07 is later). S06 has no dependencies. S07 depends on `[S01, S02]`.

Auto-mode completed S01-S04, then S07 became dispatchable. S06 (no deps) should also be dispatchable but is blocked by the positional fallback checking S05 (which is waiting on S07).

The dispatch guard was invoked 4+ times for `execute-task/M022/S06/T02` (journal shows repeated `dispatch-match` events at `2026-04-11T04:44:49`, `05:14:27`, `17:31:00`, `17:31:12`, `17:31:28`), each time returning the same blocking error.

## Forensic Evidence

- **40 anomalies** detected across the session (485 units, $1270 total)
- `[ERROR] missing-artifact`: `plan-slice/M022/S05` and `execute-task/M022/S05` recorded as completed but artifacts missing on disk
- `completed-units.json` contains malformed entry `"S)5"` (likely corruption)
- `[ERROR] doctor-issue`: Duplicate task ID T02 in S06-PLAN.md (lines 50 and 155) — separate bug but compounds the dispatch failure
- Journal shows 4 consecutive `dispatch-match` → immediate `iteration-start` cycles for `M022/S06/T02` with no `unit-end`, confirming the guard blocks before execution begins

---
*Auto-generated by `/gsd forensics`*

--- a/src/resources/extensions/gsd/dispatch-guard.ts
+++ b/src/resources/extensions/gsd/dispatch-guard.ts
@@ -107,6 +107,11 @@ export function getPriorSliceCompletionBlocker(
         // it may be a cross-milestone reference handled elsewhere.
       }
     } else {
+      const milestoneUsesExplicitDeps = slices.some((slice) => slice.depends.length > 0);
+      if (milestoneUsesExplicitDeps) {
+        return null;
+      }
+
       // Positional fallback is only a heuristic for legacy slices with no
       // declared dependencies. Skip any earlier slice that depends on the
       // target, directly or transitively, or we can deadlock a valid zero-dep

--- a/src/resources/extensions/gsd/tests/dispatch-guard.test.ts
+++ b/src/resources/extensions/gsd/tests/dispatch-guard.test.ts
@@ -172,6 +172,32 @@ test("dispatch guard ignores positionally-earlier reverse dependents for zero-de
   );
 });
 
+test("dispatch guard treats zero-dependency slices as independent when a milestone uses explicit deps (#3998)", (t) => {
+  const repo = setupRepo();
+  t.after(() => teardownRepo(repo));
+
+  mkdirSync(join(repo, ".gsd", "milestones", "M022"), { recursive: true });
+
+  insertMilestone({ id: "M022", title: "Mixed dependency milestone" });
+  insertSlice({ id: "S02", milestoneId: "M022", title: "Core A", status: "complete", depends: [], sequence: 2 });
+  insertSlice({ id: "S03", milestoneId: "M022", title: "Core B", status: "complete", depends: [], sequence: 3 });
+  insertSlice({ id: "S05", milestoneId: "M022", title: "Blocked integration", status: "pending", depends: ["S02", "S03", "S07"], sequence: 5 });
+  insertSlice({ id: "S06", milestoneId: "M022", title: "Independent zero-dep slice", status: "pending", depends: [], sequence: 6 });
+  insertSlice({ id: "S07", milestoneId: "M022", title: "Late prerequisite", status: "pending", depends: ["S02"], sequence: 7 });
+
+  writeFileSync(join(repo, ".gsd", "milestones", "M022", "M022-ROADMAP.md"), "# M022\n");
+
+  assert.equal(
+    getPriorSliceCompletionBlocker(repo, "main", "execute-task", "M022/S06/T02"),
+    null,
+  );
+
+  assert.equal(
+    getPriorSliceCompletionBlocker(repo, "main", "execute-task", "M022/S05/T01"),
+    "Cannot dispatch execute-task M022/S05/T01: dependency slice M022/S07 is not complete.",
+  );
+});
+
 test("dispatch guard allows slice with all declared dependencies complete", (t) => {
   const repo = setupRepo();
   t.after(() => teardownRepo(repo));

